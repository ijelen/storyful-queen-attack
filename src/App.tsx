import React, { useState, useRef } from "react";
import "./App.css";
import ChessBoardField from "./ChessBoardField";
import { Position } from "./ChessBoardField";
import { canAttack } from "./utils";

function App() {
  const [positionWhite, setPositionWhite] = useState<Position>([0, 0]);
  const [positionBlack, setPositionBlack] = useState<Position>([0, 0]);
  const whitesTurn = useRef(true);

  const getFields = () => {
    const fields = [];
    for (let i = 1; i < 9; i++) {
      for (let j = 1; j < 9; j++) {
        fields.push(
          <ChessBoardField
            key={`${i}${j}`}
            x={j}
            y={i}
            handleClick={handleClick}
            white={
              positionWhite[0] === j && positionWhite[1] === i ? true : false
            }
            black={
              positionBlack[0] === j && positionBlack[1] === i ? true : false
            }
          />
        );
      }
    }
    return fields;
  };

  const handleClick = (pos: Position) => {
    if (whitesTurn.current) {
      // If white clicks on the black's field, exit early
      if (pos[0] === positionBlack[0] && pos[1] === positionBlack[1]) {
        return;
      }
      setPositionWhite(pos);
      whitesTurn.current = false;
    } else {
      // If black clicks on the white's field, exit early
      if (pos[0] === positionWhite[0] && pos[1] === positionWhite[1]) {
        return;
      }
      setPositionBlack(pos);
      whitesTurn.current = true;
    }
  };

  const UIletters = ["A", "B", "C", "D", "E", "F", "G", "H"];

  return (
    <div className="ui container" style={{ marginTop: "3rem" }}>
      <h1 className="ui header">Option three: Queen Atack</h1>
      <p>
        Given the position of two queens on a chess board, indicate whether or
        not they are positioned so that they can attack each other.
      </p>
      <p>
        In the game of chess, a queen can attack pieces which are on the same
        row, column, or diagonal. A chessboard can be represented by an 8 by 8
        array.
      </p>
      <p>
        So if you're told the white queen is at c5 (zero-indexed at column 2,
        row 3) and the black queen at f2 (zero-indexed at column 5, row 6), then
        you'd know you've got a set-up like so:
      </p>
      <p>
        In this case, that answer would be yes, they can, because both pieces
        share a diagonal.
      </p>
      <p>
        A queen must be placed on a valid position on the board. Two queens
        cannot share the same position.
      </p>

      <div className="ui divider" style={{ marginBottom: "3rem" }}></div>
      <div className="ui grid container">
        <div className="sixteen wide column">
          {/* Chessboard */}
          {/* https://codepen.io/jeansarlon/pen/WpZNda */}
          <div className="wrapper">
            <div className="top">
              <div className="wrapper-inner">
                {UIletters.map((letter) => (
                  <div key={letter} className="box-inner">
                    {letter}
                  </div>
                ))}
              </div>
            </div>
            {getFields()}
            <div className="bottom">
              <div className="wrapper-inner">
                {UIletters.map((letter) => (
                  <div key={letter} className="box-inner">
                    {letter}
                  </div>
                ))}
              </div>
            </div>
          </div>
        </div>
        <div className="sixteen wide column" style={{ textAlign: "center" }}>
          {canAttack(positionWhite, positionBlack) && positionWhite[0] !== 0 ? (
            <h3>
              <i className="thumbs up outline icon blue"></i>Can attack
            </h3>
          ) : (
            <h3>
              <i className="thumbs down outline icon red"></i>Can't attack
            </h3>
          )}
        </div>
      </div>
    </div>
  );
}

export default App;
