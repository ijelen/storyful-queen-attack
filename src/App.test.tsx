import React from "react";
import { render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";

import App from "./App";

test("renders Can't attack", () => {
  render(<App />);
  const element = screen.getByText(/Can't attack/i);
  expect(element).toBeInTheDocument();
});

test("Render App", () => {
  const component = renderer.create(<App />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
