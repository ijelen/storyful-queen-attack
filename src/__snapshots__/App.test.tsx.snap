// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render App 1`] = `
<div
  className="ui container"
  style={
    Object {
      "marginTop": "3rem",
    }
  }
>
  <h1
    className="ui header"
  >
    Option three: Queen Atack
  </h1>
  <p>
    Given the position of two queens on a chess board, indicate whether or not they are positioned so that they can attack each other.
  </p>
  <p>
    In the game of chess, a queen can attack pieces which are on the same row, column, or diagonal. A chessboard can be represented by an 8 by 8 array.
  </p>
  <p>
    So if you're told the white queen is at c5 (zero-indexed at column 2, row 3) and the black queen at f2 (zero-indexed at column 5, row 6), then you'd know you've got a set-up like so:
  </p>
  <p>
    In this case, that answer would be yes, they can, because both pieces share a diagonal.
  </p>
  <p>
    A queen must be placed on a valid position on the board. Two queens cannot share the same position.
  </p>
  <div
    className="ui divider"
    style={
      Object {
        "marginBottom": "3rem",
      }
    }
  />
  <div
    className="ui grid container"
  >
    <div
      className="sixteen wide column"
    >
      <div
        className="wrapper"
      >
        <div
          className="top"
        >
          <div
            className="wrapper-inner"
          >
            <div
              className="box-inner"
            >
              A
            </div>
            <div
              className="box-inner"
            >
              B
            </div>
            <div
              className="box-inner"
            >
              C
            </div>
            <div
              className="box-inner"
            >
              D
            </div>
            <div
              className="box-inner"
            >
              E
            </div>
            <div
              className="box-inner"
            >
              F
            </div>
            <div
              className="box-inner"
            >
              G
            </div>
            <div
              className="box-inner"
            >
              H
            </div>
          </div>
        </div>
        <div
          className="box"
          id="field11"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field21"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field31"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field41"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field51"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field61"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field71"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field81"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field12"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field22"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field32"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field42"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field52"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field62"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field72"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field82"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field13"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field23"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field33"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field43"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field53"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field63"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field73"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field83"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field14"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field24"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field34"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field44"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field54"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field64"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field74"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field84"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field15"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field25"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field35"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field45"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field55"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field65"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field75"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field85"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field16"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field26"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field36"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field46"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field56"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field66"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field76"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field86"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field17"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field27"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field37"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field47"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field57"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field67"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field77"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field87"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field18"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field28"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field38"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field48"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field58"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field68"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field78"
          onClick={[Function]}
        />
        <div
          className="box"
          id="field88"
          onClick={[Function]}
        />
        <div
          className="bottom"
        >
          <div
            className="wrapper-inner"
          >
            <div
              className="box-inner"
            >
              A
            </div>
            <div
              className="box-inner"
            >
              B
            </div>
            <div
              className="box-inner"
            >
              C
            </div>
            <div
              className="box-inner"
            >
              D
            </div>
            <div
              className="box-inner"
            >
              E
            </div>
            <div
              className="box-inner"
            >
              F
            </div>
            <div
              className="box-inner"
            >
              G
            </div>
            <div
              className="box-inner"
            >
              H
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="sixteen wide column"
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      <h3>
        <i
          className="thumbs down outline icon red"
        />
        Can't attack
      </h3>
    </div>
  </div>
</div>
`;
